<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Blaser Lab Data Science Online Course • datascience.curriculum</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Blaser Lab Data Science Online Course">
<meta name="description" content="What the package does (one paragraph).">
<meta property="og:description" content="What the package does (one paragraph).">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">datascience.curriculum</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9009</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/core_1.html">Core Lesson 1:  Setting Up</a></li>
    <li><a class="dropdown-item" href="articles/core_2.html">Core Lesson 2:  R Fundamentals</a></li>
    <li><a class="dropdown-item" href="articles/core_3.html">Core Lesson 3:  Next Level R</a></li>
    <li><a class="dropdown-item" href="articles/core_4.html">Core Lesson 4:  Packages</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level3">
<h3 id="welcome-to-the-blaser-lab-computational-data-analysis-course">Welcome to the Blaser Lab Computational Data Analysis Course<a class="anchor" aria-label="anchor" href="#welcome-to-the-blaser-lab-computational-data-analysis-course"></a>
</h3>
<p>This course is designed to help you improve your data management and analysis skills. The skills you will learn can be used for everything from straightforward measurements of biological data to high-dimensional data like scRNA-seq.</p>
<p>Did you ever see a figure you made last year but can’t remember how you made it? Do you need to write a Data Sharing Plan for an NIH grant but don’t know where to begin? Understanding the principles we cover here will help you and others reproduce and extend your work.</p>
<p>This is a refresh of the 2023 Workshop. It is no longer being offered as a webinar. These materials are free to use. If you would like to collaborate and have a student who would benefit from this training, you can email <a href="mailto:bradley.blaser@osumc.edu?subject=datascience">bradley.blaser@osumc.edu</a>.</p>
</div>
<div class="section level3">
<h3 id="goals">Goals<a class="anchor" aria-label="anchor" href="#goals"></a>
</h3>
<p>The goals is for you to become comfortable with the R statistical computing language and associated computational technology:</p>
<ul>
<li>Rstudio for interacting with R</li>
<li>git for version control</li>
<li>github to share your work</li>
<li>R package development to keep your code and data separate and organized</li>
</ul>
</div>
<div class="section level3">
<h3 id="core-objectives">Core Objectives<a class="anchor" aria-label="anchor" href="#core-objectives"></a>
</h3>
<p>Once you have completed this course you should be able to:</p>
<ul>
<li>properly format data for efficient computation</li>
<li>generate a table of descriptive statistics for data from a typical biological experiment</li>
<li>perform statistical testing as appropriate</li>
<li>generate publication-quality plots using ggplot</li>
<li>generate an R data package holding processed data</li>
<li>understand the difference between raw data, processed data, and analysis code</li>
<li>use basic version control functions to track and document changes to your analysis</li>
<li>publish your code so reviewers can understand how you arrived at your results</li>
</ul>
</div>
<div class="section level3">
<h3 id="optional-modules">Optional Modules<a class="anchor" aria-label="anchor" href="#optional-modules"></a>
</h3>
<p>I have several developed software modules which will be of interest to most. They aid in the programmatic handling of data we often encounter in biology. After completing the core modules you should be comfortable with the following:</p>
<ul>
<li>scRNA-seq data processing and analysis</li>
<li>ranged genomic data (ChIPseq etc.): Trace</li>
<li>making heatmaps: SummarizedHeatmap</li>
<li>programmatic generation of genbank files: Ape</li>
<li>blinding images for unbiased quantification</li>
</ul>
</div>
<div class="section level3">
<h3 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h3>
<p>The course assumes no prior knowledge of R. It is designed for biologists with an interest in analyzing high-dimensional and/or computationally-intensive data. The only prerequisites are a basic understanding of biological experimental design (controls, biological replicates, technical replicates, etc.) and a computer.</p>
<p>Before the first class you should make sure your computer is ready to go. There are several computing options to choose from.</p>
<ul>
<li>You can use your personal computer. Most/all of the software we use in the course is available for Mac and PC and will run directly on your machine. If you own the computer, just install the programs below. If it is a lab computer, have your IT admin install the programs.</li>
<li>You may have access to a lab server or cloud service running Rstudio server. This will probably be linux-based and will run everything we will be using in the course.</li>
<li>You can use an existing academic account at the Ohio Supercomputing center or your institution’s equivalent. The cost for what we will be doing will be tiny and this has the advantage of being a computing environment you may already be familiar with and have used/will use for your work.</li>
</ul>
<p>R studio cloud (now Posit cloud) is not a great option for your academic research. It is subscription-based and the rates are exorbitant compared to what you will pay at a supercomputing center at your academic institution.</p>
<p><strong>The best computing option for this class will be what you wish to use for your own research projects.</strong></p>
<p>If you wish to work on your local computer, here are links to the programs we will be using:</p>
<ul>
<li>
<a href="https://cran.rstudio.com/" class="external-link">R</a> (critical)</li>
<li>
<a href="https://www.rstudio.com/products/rstudio/download/#download" class="external-link">Rstudio</a> (critical)</li>
<li>
<a href="http://git-scm.com/downloads" class="external-link">Git</a> (recommended)</li>
<li>
<a href="https://cran.r-project.org/bin/windows/Rtools/rtools42/rtools.html" class="external-link">Rtools</a> (Windows only)</li>
<li>
<a href="https://www.evernote.com/shard/s324/sh/4a8f8452-b624-52d9-ffb7-114d5a501e95/698b49f231396e798a93b9b9d1ebe9c5" class="external-link">Xcode command line tools</a> (Mac only)</li>
</ul>
<p>Windows and Mac users: see this <a href="https://www.evernote.com/shard/s324/sh/4a8f8452-b624-52d9-ffb7-114d5a501e95/698b49f231396e798a93b9b9d1ebe9c5" class="external-link">note</a> on installing Rtools and Xcode.</p>
<p>You should also register for a free <a href="https://github.com/" class="external-link">github</a> account. Choose a name that you would be OK with putting in a publication.</p>
</div>
<div class="section level3">
<h3 id="curriculum-structure">Curriculum structure<a class="anchor" aria-label="anchor" href="#curriculum-structure"></a>
</h3>
<p>This workshop starts from the basics and moves through somewhat advanced topics. Exercises are presented at the end of each lesson. Try to complete these before moving onto the next lesson.</p>
</div>
<div class="section level3">
<h3 id="a-note-about-ai-coding-tools">A note about AI coding tools<a class="anchor" aria-label="anchor" href="#a-note-about-ai-coding-tools"></a>
</h3>
<p>AI coding tools are a powerful way to extend your skills. I use them often. <strong>But be aware that you must absolutely know how to ask focused questions and then be able to incorporate the result into your code.</strong> As of 2026, in my experience, if you ask “too big” of a question from something like ChatGPT, you will get extremely verobse, virtually unreadable code in return. This may or may not run without errors and even if it does run, it may generate results you don’t want. If you don’t know what you are looking for then this is bad. So, ask questions that are small enough that you can understand and learn from every line that the AI produces. These small tasks are usually easier for the AI to get right as well.</p>
</div>
<div class="section level3">
<h3 id="curriculum-overview">Curriculum Overview<a class="anchor" aria-label="anchor" href="#curriculum-overview"></a>
</h3>
<p>This course covers a relatively wide range of topics which may be intimidating for new R users. The first three lectures will present some basics in using R. The last three will be more advanced.</p>
<p>Even intermediate-level users with some pre-existing experience using R will likely learn some helpful information in the early lectures.</p>
<div class="section level4">
<h4 id="core-lesson-1-setting-up-your-computing-environment">
<a href="https://blaserlab.github.io/datascience.curriculum/articles/core_1.html" class="external-link">Core Lesson 1: Setting up your computing environment</a><a class="anchor" aria-label="anchor" href="#core-lesson-1-setting-up-your-computing-environment"></a>
</h4>
<ul>
<li><a href="https://raw.githubusercontent.com/blaserlab/datascience.curriculum/main/docs/core_1.R" class="external-link">Core Lesson 1 R script</a></li>
<li>Content
<ul>
<li>R and Rstudio</li>
<li>blaseRtemplates</li>
<li>Projects</li>
<li>Packages</li>
<li>Basic Git for the single user</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="core-lesson-2-basics-of-working-with-data-in-r">
<a href="https://blaserlab.github.io/datascience.curriculum/articles/core_2.html" class="external-link">Core Lesson 2: Basics of working with data in R</a><a class="anchor" aria-label="anchor" href="#core-lesson-2-basics-of-working-with-data-in-r"></a>
</h4>
<ul>
<li><a href="https://raw.githubusercontent.com/blaserlab/datascience.curriculum/main/docs/core_2.R" class="external-link">Core Lesson 2 R script</a></li>
<li>Content
<ul>
<li>Review GIT</li>
<li>File system basics</li>
<li>Data Format</li>
<li>Input and Data Types</li>
<li>Variable Assignment</li>
<li>Data Structures</li>
<li>Subsetting Data</li>
<li>Data Operations</li>
<li>Dplyr</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="core-lesson-3-more-advanced-concepts-in-r">
<a href="https://blaserlab.github.io/datascience.curriculum/articles/core_3.html" class="external-link">Core Lesson 3: More advanced concepts in R</a><a class="anchor" aria-label="anchor" href="#core-lesson-3-more-advanced-concepts-in-r"></a>
</h4>
<ul>
<li><a href="https://raw.githubusercontent.com/blaserlab/datascience.curriculum/main/docs/core_3.R" class="external-link">Core Lesson 3 R script</a></li>
<li><a href="https://www.rstudio.com/resources/cheatsheets/" class="external-link">Cheat sheets</a></li>
<li><a href="https://r-graph-gallery.com/38-rcolorbrewers-palettes.html" class="external-link">R color brewer</a></li>
<li><a href="https://r-graphics.org/" class="external-link">R graphics cookbook</a></li>
<li>Content
<ul>
<li>Scope</li>
<li>Writing Custom Functions</li>
<li>Conditionals</li>
<li>Map/Apply</li>
<li>Plotting</li>
<li>Statistical Testing</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="core-lesson-4-building-and-using-data-packages">
<a href="https://blaserlab.github.io/datascience.curriculum/articles/core_4.html" class="external-link">Core Lesson 4: Building and using data packages</a><a class="anchor" aria-label="anchor" href="#core-lesson-4-building-and-using-data-packages"></a>
</h4>
<ul>
<li><a href="https://raw.githubusercontent.com/blaserlab/datascience.curriculum/main/docs/core_4.R" class="external-link">Core Lesson 4 R script</a></li>
<li>Content
<ul>
<li>Saving R Objects</li>
<li>Setting up the Package</li>
<li>Adding Data</li>
<li>Adding Code</li>
<li>Including Raw Data</li>
<li>Documentation</li>
<li>Finishing Up</li>
<li>Loading the Data Package</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="optional-module-1-scrna-seq-tutorial">
<a href="https://blaserlab.github.io/blaseRtools/articles/scRNAseq.html" class="external-link">Optional Module 1: scRNA-seq tutorial</a><a class="anchor" aria-label="anchor" href="#optional-module-1-scrna-seq-tutorial"></a>
</h4>
<ul>
<li>10X Datasets
<ul>
<li>You will have to register your email address to download.</li>
<li><a href="https://www.10xgenomics.com/resources/datasets/500-human-pbm-cs-3-lt-v-3-1-chromium-controller-3-1-low-6-1-0" class="external-link">Chromium controller</a></li>
<li><a href="https://www.10xgenomics.com/resources/datasets/500-human-pbm-cs-3-lt-v-3-1-chromium-x-3-1-low-6-1-0" class="external-link">Chromium X</a></li>
</ul>
</li>
<li>Content
<ul>
<li>Data Management</li>
<li>Preprocessing Data in the Cloud</li>
<li>Loading Data into the cellDataSet</li>
<li>QC</li>
<li>Merging</li>
<li>Dimension Reduction</li>
<li>Batch Correction</li>
<li>Clustering</li>
<li>Gene Modules</li>
<li>Label Transfer</li>
<li>UMAP Plots</li>
<li>Gene Bubbles</li>
<li>Differential Gene Expression</li>
<li>Differential representation</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="optional-module-2-ranged-genomic-data-tutorial">
<a href="https://blaserlab.github.io/blaseRtools/articles/trace.html" class="external-link">Optional Module 2: Ranged Genomic Data tutorial</a><a class="anchor" aria-label="anchor" href="#optional-module-2-ranged-genomic-data-tutorial"></a>
</h4>
<ul>
<li>Content
<ul>
<li>The Trace Container</li>
<li>Plotting Component Tracks
<ul>
<li>Trace data</li>
<li>Gene model</li>
<li>Peaks</li>
<li>Links</li>
<li>Axis</li>
</ul>
</li>
<li>Putting the Plot Together</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="optional-module-3-heatmap-tutorial">
<a href="https://blaserlab.github.io/blaseRtools/articles/SummarizedHeatmap.html" class="external-link">Optional Module 3: Heatmap tutorial</a><a class="anchor" aria-label="anchor" href="#optional-module-3-heatmap-tutorial"></a>
</h4>
<ul>
<li>Content
<ul>
<li>Constructig a SummarizedHeatmap</li>
<li>Plotting
<ul>
<li>Main body</li>
<li>Dendrograms</li>
<li>Row and Column Annotations</li>
<li>Highlights</li>
</ul>
</li>
<li>More customization and Variations</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="optional-module-4-ape-tutorial">
<a href="https://blaserlab.github.io/blaseRtools/articles/Ape.html" class="external-link">Optional Module 4: APE tutorial</a><a class="anchor" aria-label="anchor" href="#optional-module-4-ape-tutorial"></a>
</h4>
<ul>
<li>Content
<ul>
<li>Ape Class</li>
<li>I/O</li>
<li>Annotating TF binding sites with FIMO</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="optional-module-5-image-blinding-tutorial">
<a href="https://blaserlab.github.io/blaseRtools/articles/blind_images.html" class="external-link">Optional Module 5: Image blinding tutorial</a><a class="anchor" aria-label="anchor" href="#optional-module-5-image-blinding-tutorial"></a>
</h4>
</div>
<div class="section level4">
<h4 id="course-evaluation">Course Evaluation<a class="anchor" aria-label="anchor" href="#course-evaluation"></a>
</h4>
<ul>
<li>I appreciate feedback on things that are or are not working.</li>
<li>Please email to <a href="mailto:bradley.blaser@osumc.edu?subject=datascience%20workshop%20evaluation">bradley.blaser@osumc.edu</a>
</li>
</ul>
</div>
</div>

  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing datascience.curriculum</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Brad Blaser <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-3168-5423" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
