<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Core Lesson 4:  Packages • datascience.curriculum</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Core Lesson 4:  Packages">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">datascience.curriculum</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9009</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/core_1.html">Core Lesson 1:  Setting Up</a></li>
    <li><a class="dropdown-item" href="../articles/core_2.html">Core Lesson 2:  R Fundamentals</a></li>
    <li><a class="dropdown-item" href="../articles/core_3.html">Core Lesson 3:  Next Level R</a></li>
    <li><a class="dropdown-item" href="../articles/core_4.html">Core Lesson 4:  Packages</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Core Lesson 4:  Packages</h1>
            
      

      <div class="d-none name"><code>core_4.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This lesson is on building R data packages. This is a relatively
advanced topic, but I think it is important and can be learned quickly.
This lecture assumes you have a good working knowledge of R and your
computer file system. You do not need to be an expert in R to do
this.</p>
<p>Why would you want to build a data package? Here are the stages your
data will pass through:</p>
<ul>
<li>raw data: should be unchanged after creation. Can be entirely
contained within an R data package if it is a lightweight text file.
Larger files should be stored and backed up on a network drive.</li>
<li>processed data: these are data that have been processed into a form
that can be used for plotting or statistical analysis with only a few
lines of code and a second or two of computation time.</li>
<li>analyzed data: the final tables or figures that convey the results
of your experiment to your reader.</li>
</ul>
<p>The data package is meant to hold processed data and the code you
used to generate it from the raw data. Since these two things are
related, it is natural to keep them together in the same digital file.
Often data processing is computer resource or time-intensive and so you
don’t want to repeat it every time you interact with your work. Some
analyses (like single cell RNAseq dimension reduction) are not
absolutely reproducible in the strictest sense and so saving a processed
data object means you are always starting from the same place with your
subsequent analysis.</p>
<p>What should and shouldn’t be in a data package? This you will have to
decide for yourself. There is some overhead (effort) to making a data
package, so you won’t want to save things that can be very rapidly
calculated or derived from other data objects. Generally you want a data
package to be no more than a few GB in size so you may need to be
selective with what you save. If it takes more than 3-4 seconds to
calculate data for a table or figure, then I will usually put it in the
data package. Here are the types of things I do and don’t include</p>
<p>Do include:</p>
<ul>
<li>processed tabular data for generating statistics or plots</li>
<li>single cell data objects (e.g. cellDataSet or Seurat objects)</li>
<li>qc files for single cell analysis</li>
<li>Trace, Ape, SummarizedHeatmap objects shown in this course</li>
<li>Other bioconductor containers: SummarizedExperiment, GenomicRanges
etc.</li>
</ul>
<p>Do not include:</p>
<ul>
<li>raw sequencing data, genomic features files, or cell-barcode
matrices</li>
<li>subsets of any object already in the data package</li>
<li>plots</li>
</ul>
<p>In summary, the benefits of using a data package are:</p>
<ul>
<li>Your processed data is linked to the code used to generate it</li>
<li>All processed data and code are in a single file that can be
distributed and used by collaborators or reviewers</li>
<li>Data objects are documented so you have a better chance of recalling
what they are and how they were generated at any point in the
future</li>
<li>Data objects are version-controlled and not easy to change
inadvertently.</li>
<li>Data versions and analysis code versions are linked, so you can
always recreate the state of your project from any prior commit.</li>
</ul>
</div>
<div class="section level2">
<h2 id="saving-r-objects">Saving R objects<a class="anchor" aria-label="anchor" href="#saving-r-objects"></a>
</h2>
<p>Once you have identified a data object you would like to put into a
package, you first need to save it to disk as a <code>.rda</code> file.
Usually you are going to be starting from within an analysis project.
Here are some rules/best practices:</p>
<ul>
<li>create a new directory called “data” to hold your data object</li>
<li>save only 1 object per file. R will let you save more but this is
not a desirable way to go.</li>
<li>give the file the same name as the object.</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># some demo data to save</span></span>
<span><span class="va">demo_iris_data</span> <span class="op">&lt;-</span> <span class="va">iris</span></span>
<span></span>
<span><span class="co"># make a data directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># save the data as a .rda file</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">demo_iris_data</span>, </span>
<span>     file <span class="op">=</span> <span class="st">"data/demo_iris_data.rda"</span><span class="op">)</span></span></code></pre></div>
<p>At this point you could just stop and then read in the data later
when needed using <code>load("data/demo_iris_data.rda")</code>. But it
is better to put this into a package and then install the package in
your project for the reasons mentioned above.</p>
</div>
<div class="section level2">
<h2 id="setting-up-the-package">Setting up the package<a class="anchor" aria-label="anchor" href="#setting-up-the-package"></a>
</h2>
<p>First you might want to update your ~/.Rprofile. This will add your
contact info to the package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span></span>
<span>  usethis.full_name <span class="op">=</span> <span class="st">"Jane Doe"</span>,</span>
<span>  usethis.protocol  <span class="op">=</span> <span class="st">"ssh"</span>,</span>
<span>  usethis.description <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"Authors@R"</span> <span class="op">=</span> <span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/person.html" class="external-link">person</a></span><span class="op">(</span></span>
<span>        <span class="st">"Jane"</span>, <span class="st">"Doe"</span>,</span>
<span>        email <span class="op">=</span> <span class="st">"jane@example.com"</span>,</span>
<span>        role <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"aut"</span>, <span class="st">"cre"</span><span class="op">)</span>,</span>
<span>        comment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>ORCID <span class="op">=</span> <span class="st">"JANE'S-ORCID-ID"</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    Version <span class="op">=</span> <span class="st">"0.0.0.9000"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  usethis.destdir <span class="op">=</span> <span class="st">"~/the/place/where/I/keep/my/R/projects"</span>,</span>
<span>  usethis.overwrite <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Then use the dedicated function from blaseRtemplates to set up the
package project.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">blaseRtemplates</span><span class="fu">::</span><span class="fu"><a href="https://blaserlab.github.io/blaseRtemplates/reference/initialize_package.html" class="external-link">initialize_package</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"~/r_projects/workshop.data"</span><span class="op">)</span></span></code></pre></div>
<p>As with a regular project, you want to make a readme and set it up to
use git using the initialization script:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make a software license </span></span>
<span><span class="fu">usethis</span><span class="fu">::</span><span class="fu"><a href="https://usethis.r-lib.org/reference/licenses.html" class="external-link">use_mit_license</a></span><span class="op">(</span><span class="st">"&lt;your name here&gt;"</span><span class="op">)</span> </span>
<span><span class="co"># generate a readme file to explain your work </span></span>
<span><span class="fu">usethis</span><span class="fu">::</span><span class="fu"><a href="https://usethis.r-lib.org/reference/use_readme_rmd.html" class="external-link">use_readme_md</a></span><span class="op">(</span>open <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span>
<span><span class="co"># *** Only if developing a package *** </span></span>
<span><span class="co"># uncomment and run to generate a news file to document updates. </span></span>
<span><span class="fu">usethis</span><span class="fu">::</span><span class="fu"><a href="https://usethis.r-lib.org/reference/use_news_md.html" class="external-link">use_news_md</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="co"># set your default branch to "main" for git init </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"git config --global init.defaultBranch main"</span><span class="op">)</span> </span>
<span><span class="co"># initialize git </span></span>
<span><span class="fu">usethis</span><span class="fu">::</span><span class="fu"><a href="https://usethis.r-lib.org/reference/use_git.html" class="external-link">use_git</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="co"># initialize github </span></span>
<span><span class="fu">usethis</span><span class="fu">::</span><span class="fu"><a href="https://usethis.r-lib.org/reference/use_github.html" class="external-link">use_github</a></span><span class="op">(</span>private <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="co">### Delete this file after initializing the project! ###</span></span></code></pre></div>
<p>A package is like a project but with a few additional requirements,
like a DESCRIPTION file. The initialize_package function will set that
up for you. Here is how it will look when you are done:</p>
<pre><code>.
|-- blaser.park.datapkg.Rproj
|-- data
|   |-- cds_heme_combined.rda
|   |-- cds_heme_combined_tm.csv
|   `-- human_hsc_pseudobulk_res.rda
|-- DESCRIPTION
|-- inst
|   |-- data-raw
|   |   |-- git_commands.R
|   |   |-- hsc_pseudobulk.R
|   |   `-- reprocess_cds.R
|   `-- extdata
|       |-- cds_marrow_heme_models
|       |   |-- file_index.rds
|       |   |-- rdd_pca_transform_model.rds
|       |   |-- rdd_umap_transform_model_annoy.idx
|       |   |-- rdd_umap_transform_model.rds
|       |   `-- rdd_umap_transform_model_umap.idx
|       `-- gene_targets.csv
|-- library_catalogs
|   `-- blas02_blaser.park.datapkg.tsv
|-- LICENSE
|-- LICENSE.md
|-- man
|   |-- cds_heme_combined.Rd
|   `-- human_hsc_pseudobulk_res.Rd
|-- NAMESPACE
|-- NEWS.md
|-- R
|   `-- data.R
`-- README.md

8 directories, 23 files
</code></pre>
<p>Some important things to note</p>
<ul>
<li>DESCRIPTION: This is the file that indicates to R that this is a
package. The most important line for you to interact with is version.
Every time you make a new version of the package (add, delete or edit
data), you should increment the version number. Title and description
can be edited to provide a general description of the purpose of the
package but these are not critical. You should leave the other lines
as-is.</li>
<li>R/: This directory should contain R files. If you were building a
package with functions, this is where they would go. For a data-only
package, you will create a single R file here called “data.R”.</li>
<li>NAMESPACE: This is not used by data-only packages. Do not edit.</li>
<li>README.md and NEWS.md are text files you can edit as described
above.</li>
<li>The other files mostly boilerplate configuration files which should
be left alone.</li>
</ul>
</div>
<div class="section level2">
<h2 id="adding-data">Adding data<a class="anchor" aria-label="anchor" href="#adding-data"></a>
</h2>
<p>This is very simple. Just move the data directory from the analysis
project into the root directory of the new data package. If you are
editing an existing package, just move the new .rda files. You can only
have 1 data directory.</p>
</div>
<div class="section level2">
<h2 id="adding-data-processing-code">Adding data-processing code<a class="anchor" aria-label="anchor" href="#adding-data-processing-code"></a>
</h2>
<p>It is good to keep the processed data with the code you used to
process it. That way if there is a problem you can track it down easily.
This code should go in a new directory called “data-raw”. This is an
unfortunate name because there is no data in there, only code.</p>
<p>You need to enclose this in a directory called “inst”. Everything in
the inst directory gets installed with your package.</p>
<div class="section level3">
<h3 id="option-include-raw-data">Option: include raw data<a class="anchor" aria-label="anchor" href="#option-include-raw-data"></a>
</h3>
<p>If you want to actually include the raw data or other files with your
package, make a directory called “inst/extdata” and put them in
there.</p>
</div>
</div>
<div class="section level2">
<h2 id="documenting-data">Documenting data<a class="anchor" aria-label="anchor" href="#documenting-data"></a>
</h2>
<p>Documenting your data is useful to yourself and others.</p>
<p>All objects in your data folder need to have a documentation entry.
Documentation is done in a structured text language called Roxygen which
is very easy to work with.<br>
The easiest way to start is with annotating a simple data frame. The
<code>sinew</code> package will help you get the format correct.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># install sinew if you don't have it</span></span>
<span><span class="fu">sinew</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/sinew/man/makeOxygen.html" class="external-link">makeOxygen</a></span><span class="op">(</span><span class="va">demo_iris_data</span><span class="op">)</span></span></code></pre></div>
<p>Copy this from the console then just fill in the blanks. Usually the
most important things are to have a clear, descriptive, unique title and
then in the description line, to describe how the data was generated and
where to find the code for it. For example: This is a standard data set
distributed with R, with modifications. See
data-raw/file_with_code_you_used.R.</p>
<p>Sinew only works for data frames and functions. If you want to
document other types of objects, you can just copy/paste the format.</p>
</div>
<div class="section level2">
<h2 id="finish-up-the-data-package">Finish up the data package<a class="anchor" aria-label="anchor" href="#finish-up-the-data-package"></a>
</h2>
<p>Make sure you remember to increment the version of the data in the
DESCRIPTION file. Then jot a few notes about what changed in
NEWS.md.</p>
<p>Finally you need to tell R to generate the documentation files and
build the binary data package.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># generate the formatted documentation manuals</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/document.html" class="external-link">document</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># optionally you can now commit and push to github using the terminal</span></span>
<span></span>
<span><span class="co"># build the binary data package</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/build.html" class="external-link">build</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>That’s it. When you run the last command, a .tar.gz file will be
generated in the directory enclosing your data package code. You can
move that wherever you like. Better yet you can provide a file path
where R will save the binary package. I usually save to network storage.
You want to be sure to increment your version number each time you make
changes so you don’t overwrite old versions of your data. It is always
good to be able to go back in time if needed.</p>
<p>The package can easily be shared inside your firewall with local
collaborators. For external collaborators or reviewers, you can use
things like figshare to share large data packages. Figshare has a
straightforward web interface and the option to keep data private until
the manuscript is published.</p>
</div>
<div class="section level2">
<h2 id="a-note-about-saving-single-cell-rna-seq-data">A note about saving single cell RNA-seq data<a class="anchor" aria-label="anchor" href="#a-note-about-saving-single-cell-rna-seq-data"></a>
</h2>
<p>Single cell RNA-seq objects have three essential components: a
cell-gene matrix, cell metadata and gene metadata. Traditionally,
cell-gene matrix is held internally as a sparse matrix. This is very
efficient way to store these type of data, although going beyond about
100k cells in an object leads to extended loading times and memory
issues.</p>
<p>A better approach has been developed by Cole Trapnell using the
BPCells package. This stores the cell-gene matrix on disk and reads it
when necessary. This skips the loading time and requires a much smaller
memory footprint. Using this approach will allow you to have objects
containing millions of cells.</p>
<p>Note 1: The metadata tables are still held in memory, so for 1
million cells, your cell metadata table and other internal objects will
still have 1 million rows in them. These can take some time and
memory.</p>
<p>Note 2: Seurat objects hold multiple transformed copies of the
cell-barcode matrix, e.g. raw counts, log transformed counts, etc. This
is very storage/memory inefficient. I recommend against using this
format whenever possible for the reasons above.</p>
<p>Even though the monocle3 approach is very good, there are some
implementation issues. I have written a wrapper function to save
monocle3 objects which you should use:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">blaseRtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/blaseRtools/man/save_monocle_disk.html" class="external-link">save_monocle_disk</a></span><span class="op">(</span><span class="va">object</span>, data_directory <span class="op">=</span> <span class="st">"data"</span>, extdata_directory <span class="op">=</span> <span class="st">"inst/extdata"</span><span class="op">)</span></span></code></pre></div>
<p>This will save your object into the extdata directory, as is required
by the monocle3 format. It will put a placeholder object into the data
directory. Why?</p>
<ul>
<li>it allows you to document the object like you would any other data
object, described above</li>
<li>it prevents namespace conflicts in your package</li>
</ul>
<p>When you do this, monocle3 will put some temp directories into your
project root folder. Remember to delete those before committing to
git.</p>
<p><strong>Critical: remember to put inst/extdata into your .gitignore
file</strong> You don’t want to try to commit that directory.</p>
</div>
<div class="section level2">
<h2 id="installing-the-data-package">Installing the Data Package<a class="anchor" aria-label="anchor" href="#installing-the-data-package"></a>
</h2>
<p>The packages we make may be a little different from typical R
packages. They have no functions which is somewhat unusual but the
biggest difference is size. If you are working with single cell data, it
is likely that the size of your data package will exceed what R can
handle with its normal mechanisms for loading data.</p>
<p>Instead we use functions from a couple of packages to load the data
into your R session as a digital “pointer”. Until you ask R to reference
a particular data item, it will reside in memory as a tiny digital
address to an area on your hard drive. When you ask R to reference those
data with you code, it is then loaded into memory to be used. This has
the additional advantage of reducing memory requirements for your
work.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install the data package</span></span>
<span><span class="fu">blaseRtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/blaseRtools/man/project_data.html" class="external-link">project_data</a></span><span class="op">(</span><span class="st">"path/to/directory/containing/data"</span><span class="op">)</span></span></code></pre></div>
<p>The function will go to the directory where your data lives, check
for the latest version, and install it if necessary. This is the
recommended way to go because you almost always want the latest version
and you don’t want to waste time installing it if necessary.</p>
<p>Then the function loads the digital pointers into your R session.
Check out the environment dropdown menu to see for yourself.</p>
</div>
<div class="section level2">
<h2 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<p>By now you should be comfortable setting up an R project, analyzing
your data and making figures for a manuscript or presentation.</p>
<ul>
<li>Make an R data package with your own processed experimental data and
the processing code you used</li>
<li>Save it to a network (archival) drive</li>
<li>Load the data package back into your project using
<code>blaseRtemplates::project_data()</code>.</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
